core/src/main/java/org/apache/calcite/rel/rules/JoinConditionOrExpansionRule.java

```diff
@@ -140,30 +141,68 @@ private List<RexNode> splitCond(Join join) {
   }
 
   private boolean isValidCond(RexNode node, int leftFieldCount) {
-    if (!(node instanceof RexCall)) {
-      return false;
-    }
+    boolean hasJoinKeyCond = false;
+    List<RexNode> conds = RelOptUtil.conjunctions(node);
+    for (RexNode cond : conds) {
+      if (!(node instanceof RexCall)) {
+        return false;
+      }
 
-    RexCall call = (RexCall) node;
-    SqlKind kind = call.getKind();
-    switch (kind) {
-    case EQUALS:
-    case IS_NOT_DISTINCT_FROM:
-      RexNode left = call.getOperands().get(0);
-      RexNode right = call.getOperands().get(1);
-
-      if (left instanceof RexInputRef && right instanceof RexInputRef) {
-        RexInputRef leftRef = (RexInputRef) left;
-        RexInputRef rightRef = (RexInputRef) right;
-        return (leftRef.getIndex() < leftFieldCount && rightRef.getIndex() >= leftFieldCount)
-            || (leftRef.getIndex() >= leftFieldCount && rightRef.getIndex() < leftFieldCount);
+      RexCall call = (RexCall) cond;
+      // Checks if the call is valid for use as a join key.
+      if (isEquiJoinCond(call, leftFieldCount)) {
```

Comment:
@mihaibudiu If I understand correctly, is the logic you mentioned here?

Reply:
Related, but not the same.
You want all expressions that do NOT touch both inputs of the join.

Reply:
they can be very complex, as long as they don't look at both inputs.

Reply:
> they can be very complex, as long as they don't look at both inputs.

Oh, I understand what you mean. For example, if all the columns in the `call` come from the left input, then this call is a valid one.

Reply:
that's right

Reply:
Fixed